qemu/tests/passt_transfer_tests.py

```diff
@@ -0,0 +1,121 @@
+        if s != 0:
+            test.error("Get file md5sum failed as %s" % o)
+        return re.findall(r"\w{32}", o)[0]
+
+    timeout = params.get_numeric("login_timeout", '360')
+    receive_cmd = params.get("receive_cmd")
+    sent_cmd = params.get("sent_cmd")
+    tmp_dir = params["tmp_dir"]
+    filesize = int(params.get("filesize"))
+    dd_cmd = params["dd_cmd"]
+    file_md5_check_timeout = params.get_numeric("file_md5_check_timeout", '120')
+    fw_stop_cmd = params["fw_stop_cmd"]
+
+    sessions = {}
+    ifname = {}
+    vms = []
+
+    error_context.context("Boot vms for test", test.log.info)
+    for vm_name in params.get("vms", "avocado-vt-vm1 vm2").split():
+        vms.append(env.get_vm(vm_name))
+
+    for vm in vms:
+        vm.verify_alive()
+        sessions[vm] = vm.wait_for_serial_login(timeout=timeout)
+
+    if params.get_boolean("ipv6"):
+        gateway_address = utils_net.get_default_gateway(ip_ver='ipv6')
+        local_address = "::1"
+        ifname[vm] = utils_net.get_linux_ifname(sessions[vm],
+                                                vm.get_mac_address())
+        gateway = f"{gateway_address}%{ifname[vm]}"
+    else:
+        gateway = utils_net.get_default_gateway(ip_ver='ipv4')
+        local_address = "localhost"
+    # prepare test data
+    guest_path = (tmp_dir + "src-%s" % utils_misc.generate_random_string(8))
+    dest_path = (tmp_dir + "dst-%s" % utils_misc.generate_random_string(8))
+    host_path = os.path.join(test.tmpdir, "tmp-%s" %
+                             utils_misc.generate_random_string(8))
+    test.log.info("Test setup: Creating %dbytes file on host", filesize)
+    process.run(dd_cmd % (host_path, filesize), shell=True)
+
+    try:
+        src_md5 = (crypto.hash_file(host_path, algorithm="md5"))
+        error_context.context("md5 value of data from src: %s" % src_md5,
+                              test.log.info)
+        # transfer data from host to guest
+        host2guest_port = 10001
+        for vm in vms:
+            error_context.context("Transfer data from host to %s" % vm.name,
+                                  test.log.info)
+            sessions[vm].cmd_output_safe(fw_stop_cmd)
+            sessions[vm].sendline(receive_cmd % (host2guest_port, guest_path))
+            time.sleep(5)
+            process.run(sent_cmd % (host_path, local_address, host2guest_port), shell=True)
+            dst_md5 = get_file_md5sum(guest_path, sessions[vm],
+                                      timeout=file_md5_check_timeout)
+            error_context.context("md5 value of data in %s: %s" % (vm.name, dst_md5),
+                                  test.log.info)
+            if dst_md5 != src_md5:
+                test.fail("File changed after transfer host -> %s" % vm.name)
+            host2guest_port += 1
+
+        # transfer data between guest2guest
+        guest2guest_port = 10001
+        for vm1 in vms:
+            for vm2 in vms:
+                if vm1 != vm2:
```

Comment:
Only 2 vms, the easy way is `src, dst = vms`, then `src, dst = reversed(vms)`

Reply:
All comments has been updated, please help me review it agian if you have free time. Thanks a lot.